fix: example-value-or-externalValue too broad JsonPath expressions

m-mohr opened the pull request:
Related to #883 (partially, see comments).

**Checklist**

- [x] Tests added / updated
- [ ] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Additional context**

$..examples is too broad and incorrectly complains about examples that contain a property examples, see https://github.com/stoplightio/spectral/issues/883#issuecomment-571580864

Please some one double-check whether I found all paths for the examples in the OAS3 spec. Not sure about Swagger2?!

Edit: I'm not sure why the CI fails. Is it because some JSON path expressions return an empty array? If yes, the old JSON path would also fail on OpenAPI specs with no examples, right?

nulltoken commented:
@m-mohr Could you please add a test that would only pass with that change ("examples that contain a property examples") for instance?

nulltoken commented on src/rulesets/oas/index.json:
```diff
@@ -115,7 +115,12 @@
       "recommended": true,
       "type": "style",
-      "given": "$..examples.*",
+      "given": [
```
@m-mohr This rule currently applies to both OAS2 and OAS3.

Could you please double check that all the new paths make sense in both context (cf. https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json and https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v3.0/schema.json).

Wouldn't that be the case, an approach would be to create two new rules (prefixed `oas2-` and `oas3-`) that would only target the relevant paths considering the spec version.

nulltoken replied:
@P0lip would dropping a rule and creating two new ones be considered as a breaking change?

P0lip replied:
I'd rather avoid creating a new rule if we can help it, but if we *really* need to introduce a separate rule, that wouldn't necessarily be a breaking change.
Is there any way we can leverage multiple paths to target relevant parts of the document?

m-mohr replied:
@nulltoken Are you sure this applies to OAS2? I can't find any info on examples with value or externalValue in the OAS2 spec. 🤔

nulltoken replied:
@m-mohr Sorry, I was unclear. Let me try and reword my thoughts: The currently merged rule applies to both OAS2 and OAS3 (with the `"$..examples.*"` JsonPath expression).

I was requesting you to double check that all the new paths this PR adds also apply to OAS2 and OAS3. If that's not the case, then I fear we'll have to eventually have to end up with two rules: one for OAS2 and one for OAS3, each of them leveraging the `format` rule attribute as a target specifier.

nulltoken replied:
@m-mohr Duh. My brain just exploded for a minute. 😜 

You're right. 

Although the path applies to both specs, the rule only makes sense in OAS3 context (as hinted in #883).

How about about renaming that rule to `oas3-example-value-or-externalValue` and adding a `formats` specifier to make sure it only considers OAS3 documents?

m-mohr replied:
Sure, would make sense, but is a breaking change. So I'd probably prefer to merge this as non-breaking change to get the fix faster into a release and then change the name for the next major release.

nulltoken replied:
>So I'd probably prefer to merge this as non-breaking change to get the fix faster into a release

@m-mohr :+1: Let's deal with this later as part of #883.

m-mohr commented:
@nulltoken I could post an example here, but I doubt I can add a good test. I'm not really into your test suite and don't realky know what to do there. Could someone make a test out if a json example I post? Or would a new scenario file be enough? 🤔

nulltoken commented:
>Or would a new scenario file be enough?

@m-mohr That would be perfect!

nulltoken commented:
@m-mohr The tests are currently failing because the test documents (incorrectly) expose the `examples` at the root of the document. They were previously passing because of the (too) broad JsonPath expression.

Could you please update them so that they pass? This would require simple changes in the test documents and expectations.

>I'm not really into your test suite and don't really know what to do there. 

An example below:

```diff
diff --git a/src/rulesets/oas/__tests__/example-value-or-externalValue.ts b/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
index 796ccc0..820f547 100644
--- a/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
+++ b/src/rulesets/oas/__tests__/example-value-or-externalValue.ts
@@ -56,21 +56,21 @@ describe('example-value-or-externalValue', () => {
 
   test('multiple examples - return warnings if missing externalValue and value in one', async () => {
     const results = await s.run({
-      examples: { first: { value: 'value1' }, second: { externalValue: 'external-value2' }, third: {} },
+      components: { examples: { first: { value: 'value1' }, second: { externalValue: 'external-value2' }, third: {} } },
     });
     expect(results).toEqual([
       {
         code: 'example-value-or-externalValue',
         message: 'Example should have either a `value` or `externalValue` field.',
-        path: ['examples', 'third'],
+        path: ['components', 'examples', 'third'],
         range: {
           end: {
-            character: 15,
-            line: 8,
+            character: 17,
+            line: 9,
           },
           start: {
-            character: 12,
-            line: 8,
+            character: 14,
+            line: 9,
           },
         },
         severity: DiagnosticSeverity.Warning,
```

**Note:** in order to repro those failure locally, running `yarn test "example-value"` should work.

m-mohr commented:
@nulltoken Thanks for the hint. I updated the tests as requested and added new tests, also checking the
* other paths (parameters, headers, content)
* examples on top level
* examples properties in examples

So I think I made all your requested changes.

nulltoken commented:
@m-mohr Looks very good to me.

@P0lip How do you feel about this?

m-mohr commented:
Great!
Why's the `stoplightio.spectral` check hanging?

P0lip commented:
> Why's the stoplightio.spectral check hanging?

Most likely some permissions issue, as you don't belong to Stoplight organization.
@XVincentX any clue? It's Azure Pipelines.

P0lip commented on src/rulesets/oas/index.json:
```diff
@@ -115,7 +115,12 @@
       "type": "style",
-      "given": "$..examples.*",
+      "given": [
+        "$..content.*.examples.*",
```
Any chance we could narrow these down?
`$..` path expressions are fairly expensive, as we need to traverse deeply pretty much the whole object.
If we could have something more or less close to
`$.paths.*..parameters` and `$.components..parameters` that would be already a huge step forward.

Same for `content` and `headers`. If it happens to be very troublesome to implement, I can live with it.

m-mohr replied:
It's just a matter of going through the spec and compiling all cases. I'd expect that it blows up the list quite a lot . Also, I'd not be sure whether I found all cases and I would expect anybody to carefully double-check it. So I'll only do the work if anybody would agree on checking it.

m-mohr replied:
I just tried to resolve `"$..content.*.examples.*",` and it seems to become 
```
        "$.paths.*.*.requestBody.content.*.examples.*",
	"$.paths.*.*.callbacks.*.*.requestBody.content.*.examples.*",
        "$.components.callbacks.*.*.*.requestBody.content.*.examples.*",
        "$.components.requestBodies.*.content.*.examples.*",
```
BUT: I'm not at all sure whether that's correct, but I tried my best. And I think that was the easiest case. Also, some of the * could be replaced by (get or post or patch or put or ...), but I'm not sure how OR works in JSON Path without blowing the expression to be overly long and complicated. So I think I'd not be able to provide a proper list at the moment, time is too limited.

m-mohr replied:
I committed the version above, the tests succeed.

P0lip replied:
@m-mohr yeah, that's much better. Thanks!

XVincentX commented:
Checking this out @P0lip

XVincentX commented:
@P0lip For some reason the build was not triggered — some web hook failed. I've run this manually and it passed the tests, we can just merge it.

XVincentX commented:
Ah ok — I think the reason is that this is an external PR.
